### PR TITLE
Encapsulate logs metadata. Remove Objective-C support.

### PR DESCRIPTION
**Pull request by Henbyr**
## Description
I spent some time auditing `JustLog` and found few things we can improve / modernise. That's just a one proposal for `JustLog` improvements

## Motivation and Context
Creating new protocol `Loggable` for encapsulating log's metadata into one object `Log` should strategically enhance sending logs. E. g. if we decide to migrate from `JSONSerialization` to `Codable` that will be a way easier to achive with one object rather that separate params. As well it is up to client if they want to use default `Log` implementation or add their own by their own implementation.

Following changes discovered also a design issue with `static` properties we have in `Logger` in my PR. Those properties represents:

* Log keys `Logger` uses to format the final log message string which is sent to the destination.
* Default metadata which is assigned for each log message by default

To get rid of that I create new protocol `LogMessageFormatting` and its default implementation `JSONStringLogMessageFormatter`. This object contains all log message formatting logic. 
Additionally I add `FormatterKeysProviding` protocol which implemented by `FormatterKeys` which solves the problem with `static` properties in `Loggable`/`Log` and saves ability to override the log's keys in a proper way. I use keys object injection into the formatter, so we can remove `static` keys now.

Also following PR brings removal of Objective-C support

## How Has This Been Tested?
* No new unit tests were added
* Edit / adjusted existing unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: JustLog/Classes/Logger.swift [hunk 1]
```diff
@@ -9,24 +9,12 @@
...
-    internal struct QueuedLog {
-        let type: LogType
-        let message: String
-        let file: String
-        let function: String
-        let line: UInt
-    }
+public protocol Logging {
+       
+    func log(with log: Loggable)
```
Comment (antoniostrijdom) [inline JustLog/Classes/Logger.swift]: I'm not sure `with` is appropriate here? We're logging the `Loggable` instance right? Do we even need a label?

Comment (Henbyr) [reply JustLog/Classes/Logger.swift]: @antoniostrijdom I agree. What do you think if we remove `with` label and rename that method to 
`send(_ log: Loggable)`?
 
Example of method call
`Logger.shared.send(Log(...))`

File: Example/JustLog/AppDelegate.swift [hunk 2]
```diff
@@ -91,40 +91,39 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        Log.defaultUserInfo = ["application": "JustLog iOS Demo",
+                               "environment": "development",
+                               "session": sessionID]
         
         // custom keys
-        logger.logTypeKey = "logtype"
-        logger.appVersionKey = "app_version"
-        logger.iosVersionKey = "ios_version"
-        logger.deviceTypeKey = "ios_device"
+        LogKeys.logTypeKey = "logtype"
```
Comment (albertodebortoli) [inline Example/JustLog/AppDelegate.swift]: I see `LogKeys` is a struct with static vars.
For the usual reasons about static vars, is best to repurpose this to use instance vars.

File: JustLog/Classes/Log.swift [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    public static var iosVersionKey = "ios_version"
+    public static var lineKey = "line"
+    public static var messageKey = "message"
+    public static var metadataKey = "metadata"
+    public static var userInfoKey = "user_info"
+}
+
+public protocol Loggable {
+    
+    static var defaultUserInfo: [String : Any]? { get set }
```
Comment (albertodebortoli) [inline JustLog/Classes/Log.swift]: same here, no static vars, especially in protocols

File: JustLog/Classes/Log.swift [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    static var keys: LogKeys.Type { get set }
```
Comment (albertodebortoli) [inline JustLog/Classes/Log.swift]: this would then turn into 
```
var keys: LogKeys { get set }
```

File: JustLog/Classes/Log.swift [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    var file: String { get }
+    var function: String { get }
+    var line: UInt { get }
+    
+    func formattedMessage() -> String
+}
+
+public struct Log: Loggable {
+    
+    public static var defaultUserInfo: [String : Any]?
```
Comment (albertodebortoli) [inline JustLog/Classes/Log.swift]: More so here regarding static vars. Best in experiment with inheritance from a design perspective.

File: JustLog/Classes/Log.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        self.error = error
+        self.customData = customData
+        self.file = file
+        self.function = function
+        self.line = line
+    }
+    
+    public func formattedMessage() -> String {
+        var metadata = [String : Any]()
+        metadata[LogKeys.messageKey] = message
```
Comment (albertodebortoli) [inline JustLog/Classes/Log.swift]: see here? static vars are like singletons, don't allow testing.

**Review by albertodebortoli — CHANGES_REQUESTED** [review]

**Comment by albertodebortoli** [comment]
This PR doesn't deprecate Objective-C support, it removes it instead (to update the title and description).

**Comment by Henbyr** [comment]
> This PR doesn't deprecate Objective-C support, it removes it instead (to update the title and description).

Done

**Review by albertodebortoli — APPROVED** [review]
Thanks for the changes! Looking like a great step forward.

**Review by antoniostrijdom — APPROVED** [review]
LGTM

**Pull request merged**